File: 07trie/code/实现Trie(前缀树).java
class Trie {

    private Map<Character, Trie> children;
    private boolean flag = false;

    Trie() {
        this.children = new HashMap();
    }

    void insert(String word) {
        Trie cur = this;
        int length = word.length();
        char[] char_vec = word.toCharArray();
        for (int i = 0; i < length; i++) {
            Trie child;
            if (cur.children.get(char_vec[i]) == null) {
                child = new Trie();
                cur.children.put(char_vec[i], child);
            } else {
                child = cur.children.get(char_vec[i]);
            }
            cur = child;
        }
        cur.flag = true;
    }

    boolean search(String word) {
        Trie cur = this;
        int length = word.length();
        char[] char_vec = word.toCharArray();

        for (int i = 0; i < length; i++) {
            if (cur.children.get(char_vec[i]) != null) {
                cur = cur.children.get(char_vec[i]);
            } else {
                return false;
            }
        }

        return cur.flag;
    }

    boolean startsWith(String prefix) {
        Trie cur = this;
        int length = prefix.length();
        char[] char_vec = prefix.toCharArray();

        for (int i = 0; i < length; i++) {
            if (cur.children.get(char_vec[i]) != null) {
                cur = cur.children.get(char_vec[i]);
            } else {
                return false;
            }
        }

        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: datastructure/high_level/code/实现Trie(前缀树).cpp
class Trie {
    // https://leetcode.cn/problems/implement-trie-prefix-tree/
public:
    unordered_map<char, Trie*> children;
    bool flag;

    Trie() {
        children = unordered_map<char, Trie*>();
        flag = false;
    }
    
    void insert(string word) {
        auto res = this;
        for (auto c : word) {
            if (!res->children[c]) {
                res->children[c] = new Trie;
            }
            res = res->children[c];
        }
        res->flag = true;
    }
    
    bool search(string word) {
        auto res = this;
        for (auto c : word) {
            if (!res->children[c]) {
                return false;
            }
            res = res->children[c];
        }
        return res->flag;
    }
    
    bool startsWith(string prefix) {
        auto res = this;
        for (auto c : prefix) {
            if (!res->children[c]) {
                return false;
            }
            res = res->children[c];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */